perses/rjmc/geometry.py

```diff
@@ -669,667 +661,20 @@ def _torsion_logp(self, atom, xyz, torsion, atoms_with_positions, positions, bet
         involved_angles = self._get_valid_angles(atoms_with_positions, atom)
         involved_torsions = self._get_torsions(atoms_with_positions, atom)
         internal_coordinates = self._autograd_ctoi(xyz, positions[bond_atom.idx], positions[angle_atom.idx], positions[torsion_atom.idx])
-        if not Z:
-            p, Z = self._normalize_torsion_proposal(atom, internal_coordinates[0], internal_coordinates[1], bond_atom, angle_atom, torsion_atom, atoms_with_positions, positions, beta, n_divisions=5000)
-        ub_torsion = self._torsion_and_angle_potential(xyz, atom, positions, involved_angles, involved_torsions, beta)
-        p_torsion = np.exp(-ub_torsion) / Z
-        return p_torsion
+        p, Z, q, phis = self._normalize_torsion_proposal(atom, internal_coordinates[0], internal_coordinates[1], bond_atom, angle_atom, torsion_atom, atoms_with_positions, positions, beta, n_divisions=5000)
+        #find the phi that's closest to the internal_coordinate phi:
+        phi_idx, phi = min(enumerate(phis), key=lambda x: abs(x[1]-internal_coordinates[2]))
+        p_phi = p[phi_idx]
+        return np.log(p_phi)
 
     def _propose_torsion(self, atom, r, theta, bond_atom, angle_atom, torsion_atom, torsion, atoms_with_positions, positions, beta):
         """
         Propose a torsion angle, including energetic contributions from other torsions and angles
         """
         #first, let's get the normalizing constant of this distribution
-        p, Z = self._normalize_torsion_proposal(atom, r, theta, bond_atom, angle_atom, torsion_atom, atoms_with_positions, positions, beta, n_divisions=5000)
-        #now, use rejection sampling on the normalized distribution to randomly draw a phi
-        max_p = max(p)
-        logp = 0.0
-        phi = 0
-        accepted = False
-        while not accepted:
-            phi_samp = np.random.uniform(0.0, 2*np.pi)*units.radians
-            xyz, _ = self._autograd_itoc(bond_atom.idx, angle_atom.idx, torsion_atom.idx, r, theta, phi_samp, positions)
-            runif = np.random.uniform(0.0, max_p+1.0)
-            p_phi_samp = self._torsion_logp(atom, xyz, torsion, atoms_with_positions, positions, beta, Z)
-            if p_phi_samp > runif:
-                phi = phi_samp
-                logp = p_phi_samp
-                accepted = True
-            else:
-                continue
+        p, Z, q, phis = self._normalize_torsion_proposal(atom, r, theta, bond_atom, angle_atom, torsion_atom, atoms_with_positions, positions, beta, n_divisions=5000)
+        #choose from the set of possible torsion angles
+        phi_idx = np.random.choice(range(len(phis)), p=p)
+        logp = np.log(p[phi_idx])
```

Comment:
Since `p` might vary by orders of magnitude, would be better if you had `self._normalize_torsion_proposal()` return `log_p` directly as well?